### PR TITLE
feat(ticket#43):Create attendance report action

### PR DESCRIPTION
**Pull request by ghulamhafeez**
## Implementation
- Create Server action for monthly attendance report
- Bind attendance report table with real data

### UI
image

### CODE REVIEW

File: src/app/(app)/attendance-report/actions/attendance-report.ts [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+import { AttendancePercentage, AttendanceProgress } from "@/constant/constant";
+import { getAttendance } from "../../attendance/actions/getAttendance";
+
+export const getAttendanceReport = async () => {
+  const attendanceData = await getAttendance();
+
+  if (!attendanceData) return [];
+
+  const getStatus = (percentage: number): AttendanceProgress => {
+    if (percentage > AttendancePercentage.Excellent) {
```
Comment (muhammad-bilal-ali-khokhar) [inline src/app/(app)/attendance-report/actions/attendance-report.ts:12]: use `switch`

File: src/app/(app)/attendance-report/actions/attendance-report.ts [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+      return AttendanceProgress.Average;
+    }
+    return AttendanceProgress.low;
+  };
+
+  // Get the first day of the month and today's date
+  const today = new Date();
+  const firstDayOfMonth = new Date(today.getFullYear(), today.getMonth(), 1);
+
+  const isWithinCurrentMonth = (dateString: string): boolean => {
```
Comment (muhammad-bilal-ali-khokhar) [inline src/app/(app)/attendance-report/actions/attendance-report.ts:31]: add comment 
rename variable
assign type correctly

File: src/app/(app)/attendance-report/actions/attendance-report.ts [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+    const date = new Date(dateString);
+    return date >= firstDayOfMonth && date <= today;
+  };
+
+  const uniqueMembers = Array.from(
```
Comment (muhammad-bilal-ali-khokhar) [inline src/app/(app)/attendance-report/actions/attendance-report.ts:36]: confusing , you are mapping on leave agains memberId, is there need to do this ?

File: src/app/(app)/attendance-report/actions/attendance-report.ts [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+    const absentDays = new Set(
+      memberAttendance
+        .filter((attendance) => attendance.status === "reject")
+        .map((attendance) => attendance.created_at?.split("T")[0])
+    );
+
+    // For the time being i do this i created a separate ticket for it ans it fix in next Pr.
+    const leaveDays = new Set(
+      memberAttendance
+        .filter((attendance) => attendance.status === "pending")
```
Comment (muhammad-bilal-ali-khokhar) [inline src/app/(app)/attendance-report/actions/attendance-report.ts:62]: use enums for `approved, reject , pending`

File: src/app/(app)/attendance-report/actions/attendance-report.ts [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+  const isWithinCurrentMonth = (dateString: string): boolean => {
+    const date = new Date(dateString);
+    return date >= firstDayOfMonth && date <= today;
+  };
+
+  const uniqueMembers = Array.from(
+    new Set(attendanceData.map((attendance) => attendance.memberId))
+  );
+
+  return uniqueMembers.map((memberId) => {
```
Comment (Mubashar1009) [inline src/app/(app)/attendance-report/actions/attendance-report.ts:40]: where this will use i can not find any usage of this

Comment (ghulamhafeez) [reply src/app/(app)/attendance-report/actions/attendance-report.ts:40]: Fixed

File: src/app/(app)/attendance-report/actions/attendance-report.ts [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+    const memberAttendance = attendanceData.filter(
+      (attendance) =>
+        attendance.memberId === memberId &&
+        isWithinCurrentMonth(attendance.created_at)
+    );
+
+    const presentDays = new Set(
+      memberAttendance
+        .filter((attendance) => attendance.status === "approved")
```
[the remark this answers is not in the document]
Comment (ghulamhafeez) [reply src/app/(app)/attendance-report/actions/attendance-report.ts:49]: Yes, Fixed

File: src/app/(app)/attendance-report/actions/attendance-report.ts (hunk 4 above)
Comment (ghulamhafeez) [reply src/app/(app)/attendance-report/actions/attendance-report.ts:62]: Fixed

File: src/app/(app)/attendance-report/actions/attendance-report.ts (hunk 3 above)
Comment (ghulamhafeez) [reply src/app/(app)/attendance-report/actions/attendance-report.ts:36]: we have many objects in attendance data ,i get just unique memberId for each object

**Review by Mubashar1009 — APPROVED** [review]

File: src/app/(app)/attendance-report/actions/attendance-report.ts [hunk 7]
```diff
@@ -9,26 +13,26 @@ export const getAttendanceReport = async () => {
   if (!attendanceData) return [];
 
   const getStatus = (percentage: number): AttendanceProgress => {
-    if (percentage > AttendancePercentage.Excellent) {
-      return AttendanceProgress.Excellent;
+    switch (true) {
```
Comment (muhammad-bilal-ali-khokhar) [inline src/app/(app)/attendance-report/actions/attendance-report.ts:16]: not make sense

File: src/constant/constant.ts [hunk 8]
```diff
@@ -33,7 +33,7 @@ export enum LeaveRequestStatus {
 
 export enum AttendanceStatus {
   Pending = "pending",
-  Approve = "approve",
+  Approve = "approved",
```
Comment (muhammad-bilal-ali-khokhar) [inline src/constant/constant.ts:36]: Its a big change, dont do this

Comment (ghulamhafeez) [reply src/constant/constant.ts]: Fixed

**Pull request merged**